torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,367 @@
+        std::string((const char*)nodeAddrData.data(), nodeAddrData.size());
+    workerNameToURL_.insert({name, nodeAddrStr});
+  }
+
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
+  });
+
+  isRunning_ = true;
+}
+
+void TensorPipeAgent::onListenerAccepted(
+    const tensorpipe::Error& error,
+    std::shared_ptr<tensorpipe::Pipe>& pipe) {
+  // TODO: Handle listener error
+  if (error) {
+    LOG(WARNING) << "got error from listener: " << error.what();
+    return;
+  }
+
+  // Accept the next connection request
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
+  });
+
+  // Arm for server read
+  respond(pipe);
+}
+
+void TensorPipeAgent::pipeRead(
+    std::shared_ptr<tensorpipe::Pipe> pipe,
+    std::function<void(const tensorpipe::Error&, Message&&)> fn) {
+  pipe->readDescriptor(
+      [this, fn{std::move(fn)}, pipe](
+          const tensorpipe::Error& error, tensorpipe::Message&& tpMessage) {
+        if (error) {
+          fn(error, std::move(Message()));
+          return;
+        }
+
+        // Allocate memory and fill in pointers
+        Message rpcMessage = tensorpipeAllocateMessage(tpMessage);
+        TORCH_INTERNAL_ASSERT(
+            rpcMessage.tensors().size() == tpMessage.tensors.size(),
+            "Tensor num mismatch");
+        tpMessage.data = (uint8_t*)(rpcMessage.payload().data());
+        for (int i = 0; i < rpcMessage.tensors().size(); i++) {
+          auto& rpcTensor = rpcMessage.tensors()[i];
+          auto& tpTensor = tpMessage.tensors[i];
+          tpTensor.data = (uint8_t*)(rpcTensor.data_ptr());
+        }
+
+        pipe->read(
+            std::move(tpMessage),
+            [fn{std::move(fn)}, rpcMessage{std::move(rpcMessage)}](
+                const tensorpipe::Error& error,
+                tensorpipe::Message&& tpMessage) mutable {
+              fn(error, std::move(rpcMessage));
+            });
+      });
+}
+
+void TensorPipeAgent::pipeWrite(
+    std::shared_ptr<tensorpipe::Pipe> pipe,
+    Message&& rpcMessage,
+    std::function<void(const tensorpipe::Error&)> fn) {
+  tensorpipe::Message tpMessage = tensorpipeSerialize(rpcMessage);
+  pipe->write(
+      std::move(tpMessage),
+      // Note: keep payload and tensors of rpcMessage alive.
+      [rpcMessage{std::move(rpcMessage)}, fn{std::move(fn)}](
+          const tensorpipe::Error& error, tensorpipe::Message&& tpMessage) {
+        fn(error);
+      });
+}
+
+void TensorPipeAgent::sendCompletedResponseMessage(
+    std::shared_ptr<tensorpipe::Pipe>& pipe,
+    std::shared_ptr<FutureMessage>& futureResponseMessage,
+    const c10::optional<utils::FutureError>& error,
+    uint64_t messageId) {
+  if (!isRunning_) {
+    return;
+  }
+
```

Comment:
Looks like this function can run on a separate thread? If so, will it be a problem if `isRunning_ ` is set to False (shutdown) after the above check?